tracer/src/Datadog.Trace/Agent/MessagePack/SpanModelBuilder.cs

```diff
@@ -0,0 +1,149 @@
+﻿// <copyright file="SpanModelBuilder.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+namespace Datadog.Trace.Agent.MessagePack;
+
+internal struct SpanModelBuilder
+{
+    private readonly TraceChunkModel _traceChunk;
+
+    public readonly ulong LocalTraceRootSpanId;
+    public readonly bool LocalRootExists;
+
+    // for small trace chunks, use the ArraySegment<Span> copy directly, no heap allocations
+    private readonly ArraySegment<Span> _spans;
+
+    // for large trace chunks, use a HashSet<ulong> instead of iterating the array.
+    // note that his field is NOT readonly,
+    // but each copy of the same SpanModelBuilder can initialize it if needed.
+    private HashSet<ulong>? _hashSet;
+
+    public SpanModelBuilder(in TraceChunkModel traceChunk)
+    {
+        _traceChunk = traceChunk;
+        _spans = traceChunk.Spans;
+        _hashSet = null;
+
+        LocalTraceRootSpanId = traceChunk.LocalRootSpan?.SpanId ?? 0;
+        LocalRootExists = false;
+
+        // find the local root span to optimize for the common case of a "two-level" trace,
+        // i.e. a trace where all spans are direct descendants of the local root span, so there are no grandchildren,
+        // this makes it very fast to determine if a span's parent (the root) is present in the same trace chunk
+        if (LocalTraceRootSpanId > 0)
+        {
+            // the local root span is the almost always the
+            // last span in the chunk, so iterate backwards
+            for (var i = _spans.Offset + _spans.Count - 1; i >= 0; i--)
+            {
+                var span = _spans.Array![i + _spans.Offset];
+
+                if (span.SpanId == LocalTraceRootSpanId)
+                {
+                    LocalRootExists = true;
+                    break;
+                }
+            }
+        }
+    }
+
+    public readonly bool HashSetCreated => _hashSet is not null;
+
+    public bool ParentExistsForSpanAtIndex(int spanIndex)
```

Comment:
Same here but clearly more debatable, I assume one could argue that this should be part of `TraceChunkModel`.
Have you considered hiding the `ArraySegment<Span>` in `TraceChunkModel` behind a more generic structure? That way, when creating the TraceChunkModel you could keep the ArraySegment or directly store the HashSet and drop the ref to the ArraySegment (you could even directly switch to SpanModel). 

This would put a bit more pressure on the app threads and less on the AgentWriter one, but not sure it's a big deal. We would keep a bit more memory a tad longer, but not sure it's a big deal either. But again, I would understand you disagree with this one.